[SwiftSyntax] Build a static verison of the parser lib on Darwin

ahoppen opened the pull request:
As a first step to link SwiftSyntax statically against the parser library, build a static version of the parser library as part of a toolchain build. Unfortunately CMake doesn't support building static libraries that contain all transitive dependencies. Thus, I’m applyin the following workaround
- Request a build of a shared library (aka. dylib) from CMake. This makes sure that all transitive dependencies are included in the linker invocation.
-  Pass `-r` to the linker to produce a relocatable object file instead of a dylib
- Run `libtool` to generate a static library from the relocatable object file, which we install into the toolchain.

### Resulting binary size

The resulting static library is rather large (~150MB), so I went ahead and compared the size of the resulting `lit-test-helper` binary in the SwiftSyntax project to the binary sizes produced by the static parser library vendored by https://github.com/keith/StaticInternalSwiftSyntaxParser/releases/tag/5.6. The results are quite interesting



  | @keith’s static library | This PR
-- | -- | --
x86 | 56MB | 22MB
x86 + dead strip | 43MB | 22MB
arm64 | 51MB | 20MB
arm64 + dead strip | 39MB | 20MB
fat | 107MB | 111MB
fat + dead strip | 81MB | 86MB

The interesting takeaways are
- When building for a single arch, dead stripping seems to not be required anymore since this PR distributes a proper static library and not a relocatable object file 🎉 
- The fat binaries are bigger than the two single arch binaries combined. I suspect this is a bug in SwiftPM and filed https://github.com/apple/swift-package-manager/issues/5557

ahoppen commented:
@swift-ci Please build toolchain macOS

ahoppen commented:
@swift-ci Please build toolchain

keith commented:
I'm a bit surprised by the size differences here, maybe the comparison points were not on equal footing some how? Embedding a relocatable object file in a static library shouldn't be any different than linking a relocatable object file directly as far as I understand it, since once the linker requires any symbol from the object it will be loaded either way. Either way knowing it's not worse is good enough for me for this

keith commented on tools/libSwiftSyntaxParser/CMakeLists.txt:
```diff
@@ -1,61 +1,112 @@
+    #  * Pass -r to the linker to produce a relocatable object file instead of a dylib
+    #  * Run libtool (below) to generate a static library from the relocatable object file, which we install into the toolchain.
+    
+    #  Explicitly link against C(++) libraries to make sure they end up in the final relocatable object file
```
nit: we might want to word this to make it more clear that the point is to propagate these for dependees so they don't have to manually pass them, maybe something like:

```suggestion
    #  Add autolinking directives for C(++) libraries to avoid consumers having to add them manually
```

keith commented on tools/libSwiftSyntaxParser/CMakeLists.txt:
```diff
@@ -1,61 +1,112 @@
+    
+    #  Explicitly link against C(++) libraries to make sure they end up in the final relocatable object file
+    target_compile_options("${target_name}" PRIVATE
+      "SHELL:-Xclang --linker-option=-lc++"
+      "SHELL:-Xclang --linker-option=-lncurses"
+      "SHELL:-Xclang --linker-option=-lz"
```
```suggestion
      "SHELL:-Xclang --dependent-lib=c++"
      "SHELL:-Xclang --dependent-lib=ncurses"
      "SHELL:-Xclang --dependent-lib=z"
```

This is a bit more clear semantically IMO, but otherwise is the same

keith commented on tools/libSwiftSyntaxParser/CMakeLists.txt:
```diff
@@ -1,61 +1,112 @@
+
+if("${CMAKE_SYSTEM_NAME}" STREQUAL "Darwin")
+  # We only need the static library to distribute it as an xcframework for SwiftSyntax.
+  # xcframework dependencies in SwiftPM are only supported on Darwin at the moment, so there's no need to build a static library for other platforms at the moment.
```
This might be a bit misleading since other platforms could still benefit from this just not directly with swiftpm including it, but either way hopefully it's something we can do in the short term

ahoppen replied:
Reworded
```
  # TODO: Building the parser library as a static library has only been tested on Darwin so far because we only need the static library 
  # to distribute it as an xcframework for SwiftSyntax and xcframework dependencies in SwiftPM are only supported on Darwin at the moment.
  # Building and static linking the parser library on Linux might be possible, trivial.
```

keith reviewed (approved):
thanks for getting to the bottom of all of this!

ahoppen committed:
[build-script] Build internalSwiftSyntaxParser as a static library on Darwin

As a first step to link SwiftSyntax statically against the parser library, build a static version of the parser library as part of a toolchain build. Unfortunately CMake doesn't support building static libraries that contain all transitive dependencies. Thus, I’m applyin the following workaround
- Request a build of a shared library (aka. dylib) from CMake. This makes sure that all transitive dependencies are included in the linker invocation.
-  Pass `-r` to the linker to produce a relocatable object file instead of a dylib
- Run `libtool` to generate a static library from the relocatable object file, which we install into the toolchain.

Co-authored-by: Keith Smiley <keithbsmiley@gmail.com>

ahoppen commented:
@swift-ci Please build toolchain macOS

keith commented:
@ahoppen any plans to push more on this one?

cynicer commented:
Would appreciate a push on that one.

keith commented:
Side stepping the problem will also be a nice solution! https://forums.swift.org/t/a-new-swift-parser-for-swiftsyntax/59813

ahoppen commented:
Yes, that’s the reason I didn’t reply to this thread last week 😆

ahoppen commented:
Given that we are re-implementing the parse in Swift, this will not be an issue in the future, so I’m not pursuing it further.
